market/retrieval/piecehandler.go

```diff
@@ -58,19 +60,46 @@ func (rp *Provider) handleByPieceCid(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	setHeaders(w, pieceCid)
+	buf := make([]byte, 512)
+	n, _ := reader.Read(buf)
+	contentType := http.DetectContentType(buf[:n])
+
+	// rewind reader before sending
+	_, err = reader.Seek(0, io.SeekStart)
+	if err != nil {
+		log.Errorf("error rewinding reader for piece CID %s: %s", pieceCid, err)
+		w.WriteHeader(http.StatusInternalServerError)
+		stats.Record(ctx, remoteblockstore.HttpPieceByCid500ResponseCount.M(1))
+		return
+	}
+
+	setHeaders(w, pieceCid, pdpPiece, contentType)
 	serveContent(w, r, size, reader)
 
 	stats.Record(ctx, remoteblockstore.HttpPieceByCid200ResponseCount.M(1))
 	stats.Record(ctx, remoteblockstore.HttpPieceByCidRequestDuration.M(float64(time.Since(startTime).Milliseconds())))
 }
 
-func setHeaders(w http.ResponseWriter, pieceCid cid.Cid) {
+func setHeaders(w http.ResponseWriter, pieceCid cid.Cid, pdpPiece bool, contentType string) {
 	w.Header().Set("Vary", "Accept-Encoding")
-	etag := `"` + pieceCid.String() + `.gz"` // must be quoted
-	w.Header().Set("Etag", etag)
-	w.Header().Set("Content-Type", "application/piece")
 	w.Header().Set("Cache-Control", "public, max-age=29030400, immutable")
+	if pdpPiece {
```

Comment:
I wouldn't differentiate, piece is a piece.

Thinking in market v2 terms:
* Pieces are mostly separate to how they are proven on-chain
* Piece owners will be able to define some policy like ACL on a piece, but it could also include content-type settings
* Who a piece-owner is is a market v2 concern, so for now here we just have a single common default, always probing type.

Unless something would break, then this would make more sense.

Reply:
That makes sense. I will update the PR once testing for browser rendering is done